woocommerce/woocommerce-gutenberg-products-block
# Introduce pluginArea in Cart and Checkout

This PR does two things:
- Introduces `PluginArea` in Cart and Checkout, the position inside the block doesn't matter as along as it's inside any context, it also wraps them in an BlockErrorBoundary.
- Modifies `BlockErrorBoundary` to accept a `render` function, we use it to render nothing so errors are not visible for guests/shoppers.

The ErrorBoundary is limited in that it protects the block from a broken integration but doesn't protect other plugins from a broken plugin, such protection should be done on an individual level, We already do that on SlotFills (see https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3361).
Such a thing needs to be done on an individual level via a render prop and `PluginArea` doesn't offer this flexibility but I will push a PR to Gutenberg to add this.

Also, because of the way `PluginArea` works, all registered plugins would get rendered in our Area, even ones that are not entered to be shown in Cart or Checkout, anyone misbehaving there would break the rest as well. I pushed a PR to protect against that https://github.com/WordPress/gutenberg/pull/27438 and that's also pending review.

I didn't see any value in adding an abstraction layer above `registerPlugin` because it provides everything we need, however if we ever end up using the latest version of `@wordpress/plugins` instead of the `wp.plugins` one, we lose access to scope and would need to publish our own `registerPlugin` abstraction.


### Testing instructions 
1- In any plugin that loads a JS file in Cart or Checkout, add this code.
If you have WC Subscriptions, add this code to `assets/js/frontend/wcs-cart.js`
```js
const el = wp.element.createElement;
const registerPlugin = wp.plugins.registerPlugin;
registerPlugin("my-plugin", {
  render: () => {
    console.log("I'm loading!");
    return el("p", {}, "This won't get rendered");
  },
});
```
2- In the file enqueue folder (in the above case woocommerce-subscriptions.php line 378), add `"wp-element", "wp-plugins"` as dependencies.
3- Visit Cart block, you should see "I'm loading!" in console log.
4- Edit the rendered element so that instead of `"p"`, you call an undefined variable.
5- Visit Cart block again, if you're admin, you should see an error, if you're a guest, you should see nothing.


Fixes #3442
